cmd/snap/cmd_run.go

```diff
@@ -679,21 +679,31 @@ func (x *cmdRun) runCmdUnderStrace(origCmd, env []string) error {
 
 func (x *cmdRun) runSnapConfine(info *snap.Info, securityTag, snapApp, hook string, args []string) error {
 	snapConfine := filepath.Join(dirs.DistroLibExecDir, "snap-confine")
-	// if we re-exec, we must run the snap-confine from the core snap
+	// if we re-exec, we must run the snap-confine from the core/snapd snap
 	// as well, if they get out of sync, havoc will happen
 	if isReexeced() {
-		// run snap-confine from the core snap. that will work because
-		// snap-confine on the core snap is mostly statically linked
-		// (except libudev and libc)
-		snapConfine = filepath.Join(dirs.SnapMountDir, "core/current", dirs.CoreLibExecDir, "snap-confine")
+		exe, err := osReadlink("/proc/self/exe")
+		if err != nil {
+			return err
+		}
+		// snapBase will be "/snap/{core,snapd}/$rev/" because
+		// the snap binary is always under usr/bin/snap
+		snapBase, err := filepath.Abs(filepath.Join(filepath.Dir(exe), "..", ".."))
+		if err != nil {
+			return err
+		}
+		// Run snap-confine from the core/snapd snap. That
+		// will work because snap-confine on the core/snapd snap is
+		// mostly statically linked (except libudev and libc)
+		snapConfine = filepath.Join(snapBase, dirs.CoreLibExecDir, "snap-confine")
```

Comment:
This is fine, CoreLibExecDir matches the layout of the snapd/core snap regardless of distribution layout.